### ISSUE TITLE
Importing dotenv in ES6

### ISSUE THREAD

**Issue by Sparragus** [motdotla/dotenv#89]
Support on ES6 has been tricky for me. Yesterday I tried preloading dotenv with `iojs -r dotenv/config index.js`. However, it didn't work and I couldn't understand why. My app did nothing. After running the app, a second or two later it would finish without doing anything.

I ended up with a very simple way to use dotenv. I simply import it like this: 
```es6
import {} from 'dotenv/config'
import somethingElse from 'somethingElse'
...
[the rest of your code]
```

This works because of how ES6 modules imports modules. Before doing anything else in a file (say index.js) it does all the imports first. It does a depth first traversal of these imports, executing any code inside it. If we import dotenv first, it will execute config and add any env variables before doing anything else in the code. 

From [ES6 In Depth: Modules](https://hacks.mozilla.org/2015/08/es6-in-depth-modules/):
> When you run a module containing an import declaration, the modules it imports are loaded first, then each module body is executed in a depth-first traversal of the dependency graph, avoiding cycles by skipping anything already executed.

So, importing dotenv on the first line of a bootstrap file in an app will set the env vars for anything that might use them.

I suggest we add a brief section about importing dotenv with ES6 modules to the README. But before sending a pull request and wanted to open it up for discussion here.

Any thoughts?

**Comment by motdotla** [comment]
@jcblw you've done some work with ES6 right?

**Comment by maxbeatty** [comment]
@Sparragus can you provide a little more code so we can try to reproduce? I've added some examples of how to use dotenv in various ways in #90 including ES6 preloading. Are you using babel or something else for your `import` syntax? I don't believe the latest node v4.0.0 supports that out of the box yet.

I would caution against using `dotenv/config` in any context but preloading because it'll be doing additional work parsing command line arguments. `import dotenv from 'dotenv'` should do the trick

**Comment by jcblw** [comment]
I believe since we are not exporting out a *key* for` dotenv` the default import behavior of `babel` @maxbeatty mentions should work.

```javascript
import dotenv from 'dotenv'
dotenv.config()
```

I would also be interested in how your running into an issue with preloading the module. Can you post the version of node/iojs and dotenv that you are using?

**Comment by Sparragus** [comment]
@maxbeatty and @jcblw 

I've uploaded a [repo with sample code](https://github.com/Sparragus/dotenv-es6-debugging) showing everything I mentioned on my first message. 

All examples use `babel-node@5.8.23`, `dotenv@1.2.0`, and `iojs@2.2.1`.

Clone it and then make sure you also pull all the branches. Here's a [one-liner](http://stackoverflow.com/a/4754797/63825) to do that.

Each branch has a README.md with details about that particular branch. Make sure to check it out.

Start by trying out the `master` branch. Install all the dependencies with `npm install`. Then run the app with `npm start`.

Then try the other two branches: `preload-dotenv` and `import-dotenv-then-config`.

Let me know how it goes.

**Comment by maxbeatty** [comment]
`babel-node` has a `-r` option which conflicts with node's option of the same name. If you move the argument to after the script name, `log` outputs the first line but the second which depends on environment variables is undefined. If you change your start script to use `babel/register`, everything works as expected.

```
node -r dotenv/config -r babel/register index.js
```

This is an issue with using `babel-node`. Because it's primarily a development tool and not something that would be used in production, I'm inclined to close this issue.

**Comment by Sparragus** [comment]
@maxbeatty I agree with you. I didn't know about `babel-node` providing a different `-r` than node. I also liked that run script better than the one I'm using. I'm using it and it works.

Thanks all for your comments.

**Comment by colllin** [comment]
On newer version of babel (v6.x?), this command worked for me (`babel-core/register` instead of `babel/register`):

`node -r dotenv/config -r babel-core/register index.js`

Or using nodemon...

`nodemon index.js --exec 'node -r dotenv/config -r babel-core/register'`

**Comment by kachkaev** [comment]
`node --require 'dotenv/config' --require 'babel-register' index.js` worked perfectly for me, thanks @colllin!

Just a small note for those want to support the development on Windows: you should remove quotes around the module names:

```bash
node --require dotenv/config --require babel-register index.js
nodemon index.js --exec "node --require dotenv/config --require babel-register"
```

**Comment by arvi** [comment]
@Sparragus Thanks a lot. I was getting undefined on my dotenv config and changing it to `import {} from 'dotenv/config';` worked for me. :smile:

**Comment by slorenzo** [comment]
I have the following lines on `index.js`:

`import dotenv from 'dotenv';`
`dotenv.config({ path: path.join(__dirname, '.env') });`

And `process.env.MY_ENV_VAR` is a string.
But if I ask this in another file, `process.env.MY_ENV_VAR` is `undefined`
If I adding `import {} from 'dotenv/config'` to this file, `process.env.MY_ENV_VAR` now is a string.

It's really weird, but works.
Thanks @Sparragus .

**Comment by uds214125** [comment]
You should follow this step as also mentioned in dotenv documentation. 
https://github.com/motdotla/dotenv/issues/133

BTW ,  I tired it, its working fine in all the files.
0. in package.json 
   "start": "nodemon --require dotenv/config app.js --exec babel-register"

1. create config.js or env.js with these contents
  import dotenv from 'dotenv'
  dotenv.config({ silent: true });

2. in server.js/app.js or your root file. 
     import _ from './config.js';

3. check in another file 
  console.log('env : ', process.env);

**Comment by djmattski** [comment]
For anyone using Bable - this saved me: https://www.npmjs.com/package/babel-plugin-dotenv

**Comment by onwuvic** [comment]
@uds214125 Your method works fine for me. But I didn't have to preload the 'dotenv/config' on my script.

**Comment by jmbl1685** [comment]
`import { config } from 'dotenv'
config()`

**Comment by jmbl1685** [comment]
```js
import { config } from 'dotenv'
config()
```

**Comment by obscure18** [comment]
@slorenzo Since I had the same weird behaviour as you, I found out it's because, as it has been quoted above : 
> When you run a module containing an import declaration, the modules it imports are loaded first, then each module body is executed in a depth-first traversal of the dependency graph, avoiding cycles by skipping anything already executed.  

It means that the imports are executed first, so executes `import dotenv from 'dotenv';` which is simply importing `node_module/dotenv` and all other imports including your other file that **won't** have the custom .env data.  
**Then** it would execute `dotenv.config({ path: path.join(__dirname, '.env') });` which fetchs the custom .env data.  
  
When you put `import {} from 'dotenv/config'`, it would immediately import the .env data.  

To summarize :
1-` import dotenv from 'dotenv'` is executed first (if you put it before all imports, of course)
2- all other imports are made
3- fetch .env with `dotenv.config({ path: path.join(__dirname, '.env') });`

**Comment by bflemi3** [comment]
does this not work for you?

```javascript
import 'dotenv/config`
```

**Comment by Alexzanderk** [comment]
> does this not work for you?
> 
> ```js
> import 'dotenv/config`
> ```

good job the best way to import, I don't know why on the top import to a variable!

**Comment by izznogooood** [comment]
> > does this not work for you?
> > ```js
> > import 'dotenv/config`
> > ```
> 
> good job the best way to import, I don't know why on the top import to a variable!

Because this does not work in node v12.10.0
```bash
(node:19322) ExperimentalWarning: The ESM module loader is experimental.
internal/modules/esm/default_resolve.js:79
  let url = moduleWrapResolve(specifier, parentURL);
            ^

Error: Cannot find module /home/anders/dev/kaboom/node_modules/dotenv/config imported from /home/anders/dev/myfamily-api/src/index.mjs
    at Loader.resolve [as _resolve] (internal/modules/esm/default_resolve.js:79:13)
    at Loader.resolve (internal/modules/esm/loader.js:73:33)
    at Loader.getModuleJob (internal/modules/esm/loader.js:152:40)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:43:40)
    at link (internal/modules/esm/module_job.js:42:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

**Comment by webhacking** [comment]
```
import * as dotenv from 'dotenv';
dotenv.config()
```

**Comment by Weffe** [comment]
Just in case anyone is using [esm](https://github.com/standard-things/esm) to get es module syntax, then you can combine it with dotenv like so:

```
node -r esm -r dotenv/config main.js
```

This preloads your dotenv values and ensures that your references to them in your es modules will work properly.

**Comment by JNaeemGitonga** [comment]
```
import { config } from 'dotenv';
config();
```

**Comment by alex-r89** [comment]
> import { config } from 'dotenv';
> config();

The result of this is the env variables inside a parsed parent object, why is this?

IE if I console.log(config()) I get:

```
{
  parsed: {
    NODE_ENV: 'dev',
    DB_NAME: 'default',
    DB_TYPE: 'postgres',
    DB_HOST: 'localhost',
    ...rest of my .env file contents
  }
}
```

**Comment by rthadr** [comment]
Node 12.16.0 this worked for me:

`import 'dotenv/config.js';`

**Comment by Szandor72** [comment]
> Node 12.16.0 this worked for me:
> 
> `import 'dotenv/config.js';`

Confirmed for node 13.10.1. Kudos

**Comment by vborodulin** [comment]
> > Node 12.16.0 this worked for me:
> > `import 'dotenv/config.js';`
> 
> Confirmed for node 13.10.1. Kudos

works from node 10

**Comment by ajaysbugatti** [comment]
i SOLVED USING CHANGED THE DIRECTORY OF .env from root to server folder

**Comment by pxpcandy** [comment]
`import { config } from 'dotenv';` should work

**Comment by sisou** [comment]
> `import { config } from 'dotenv';` should work

This method only works from Node v14 onwards. Before that, e.g. Node 12/13, this throws an error:
```
import { config } from 'dotenv'
         ^^^^^^
SyntaxError: The requested module 'dotenv' does not provide an export named 'config'
    at ModuleJob._instantiate (internal/modules/esm/module_job.js:92:21)
    at async ModuleJob.run (internal/modules/esm/module_job.js:107:20)
    at async Loader.import (internal/modules/esm/loader.js:179:24)
```

On those older Node versions, you need to
```js
import dotenv from 'dotenv'
```

**Comment by quydox** [comment]
> @Sparragus Thanks a lot. I was getting undefined on my dotenv config and changing it to `import {} from 'dotenv/config';` worked for me.

Cool, thanks for your solution

**Comment by HappyCuppy** [comment]
None of of the suggested options here worked for me. I'm running a fresh installation of Node 16.3.0 on a fresh Manjaro installation (over WSL).
I can import `dotenv`, import `*`, `config`, etc. as suggested just fine - it's just that `process.env` ends up not including my environment variables defined in an `.env` file at the root of the project - same thing if I provide an explicit path to it in the config.

**Comment by CherryDT** [comment]
Did you do `import 'dotenv/config'? It won't work to do `import dotenv from dotenv` followed by `dotenv.config()` because then any other imported modules won't yet see the environment variables due to `import` essentially being hoisted.

**Comment by SrBrahma** [comment]
> It won't work to do `import dotenv from dotenv` followed by `dotenv.config()` because then any other imported modules won't yet see the environment variables due to `import` essentially being hoisted.

Any reference about it? I am struggling to understand this. This wasn't a thing before ESM

**Comment by CherryDT** [comment]
@SrBrahma This explains it all very well: https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/

Then think about it again - If you have files like this:

```
// main.js
import dotenv from dotenv
dotenv.config()

console.log('MAIN', process.env.STUFF) 

import 'sub.js'
```
```
// sub.js
console.log('SUB', process.env.STUFF)
```
... and you run `STUFF=abc node main.js`, then you will get this:

```
SUB undefined
MAIN abc
```

Remember this sentence from the article on the section about evaluation:

> Just as with instantiation, this is done as a depth first post-order traversal. 
Because for any module before itself is evaluated, any imported modules from top to bottom are evaluated (unless they already were evaluated elsewhere before), the imported module `dotenv` is evaluated first, and then `sub` is evaluated, and oops, `dotenv.config()` didn't run yet. Afterwards, now that all imported modules of `main` were evaluated, `main` itself can be evaluated, and `dotenv.config()` is called, but for the poor `sub` module that's already too late.

**Comment by AntonioGomes42** [comment]
I made a project, to show how it worked for me. https://github.com/AntonioGomes42/DotEnv-with-ES6-COMMOM-JS- . It was the only way i could find to make dotenv work with '.env' file an so. I hope it could help someone!

**Comment by MackKaputo** [comment]
import dotenv from "dotenv"
import path from "path"
const __dirname = path.resolve()
dotenv.config({ path: path.join(__dirname, './.env') })

**Comment by CherryDT** [comment]
@MackKaputo This has the same problem as explained above unless you put it into a separate module only for this purpose and then import _that_ module at the top of your main module.

**Comment by Naibeye** [comment]
> For anyone using Babel - this saved me: https://www.npmjs.com/package/babel-plugin-dotenv

It is worked

**Comment by motdotla** [comment]
Hi everyone. We've updated the docs to clarify this for ESM. It is right at the top fold now:

https://github.com/motdotla/dotenv#2-as-early-as-possible-in-your-application-import-and-configure-dotenv

<img width="861" alt="Screen Shot 2022-01-16 at 1 09 49 AM" src="https://user-images.githubusercontent.com/3848/149653962-82274de0-ebc3-4384-82d8-dc1b52cd0053.png">

**Comment by Hyzyr** [comment]
import dotenv from 'dotenv';

dotenv.config({path: 'config.env'});

**Comment by jeanbmar** [comment]
> Hi everyone. We've updated the docs to clarify this for ESM. It is right at the top fold now:
> 
> https://github.com/motdotla/dotenv#2-as-early-as-possible-in-your-application-import-and-configure-dotenv
> 
> <img alt="Screen Shot 2022-01-16 at 1 09 49 AM" width="861" src="https://user-images.githubusercontent.com/3848/149653962-82274de0-ebc3-4384-82d8-dc1b52cd0053.png">

How do you deal with custom path with this?

**Comment by ashen780** [comment]
`import 'dotenv/config';`

works for node version v16.14.2

**Comment by sergicanals** [comment]
import 'dotenv/config'

**Comment by alexcroox** [comment]
All these examples that work across files don't seem to show how to set the config, e.g a custom path.

**Comment by joeflan** [comment]
@alexcroox  One thing that worked for me (but I don't know if it's completely correct) to get the path working was:

```
import * as dotenv from 'dotenv'
dotenv.config({path:"custom/path/filename"})
```

**Comment by motdotla** [comment]
@alexcroox good point!

Yes, @joeflan is correct here. I'll update the README to communicate that better.

**Comment by bitdom8** [comment]
Path is default to ./root/.

To add path to current directory (Dir), we need this code inside of the file that we want all variables to be imported

```
import dotenv from 'dotenv'
import path from 'path'

import {fileURLToPath} from 'url';
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
dotenv.config({ path: path.join(__dirname, '.env') });
// const express = require('express')
import express from 'express';
```

server.js and .env can should be in the same directory if you don't manipulate the `path.join(__dirname`. 

You can reach the variables process.env.myvariable. Just be sure it works and print all variables: console.log(process.env)

Thanks

**Comment by jarodsim** [comment]
import { config } from 'dotenv'

**Comment by joshh71390** [comment]
now sure why so much stress import {config} from 'dotenv"

config()

**Comment by CherryDT** [comment]
> now sure why so much stress import {config} from 'dotenv"
> 
> config()

This will lead to surprises depending on how some modules import other modules, for the reasons I explained here: https://github.com/motdotla/dotenv/issues/89#issuecomment-974511109

**Comment by hsuabina** [comment]
> @SrBrahma This explains it all very well: https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/
> 
> Then think about it again - If you have files like this:
> 
> ```
> // main.js
> import dotenv from 'dotenv'
> dotenv.config()
> 
> console.log('MAIN', process.env.STUFF) 
> 
> import './sub.js'
> ```
> 
> ```
> // sub.js
> console.log('SUB', process.env.STUFF)
> ```
> 
> ... and you run `STUFF=abc node main.js`, then you will get this:
> 
> ```
> SUB undefined
> MAIN abc
> ```
> 
> Remember this sentence from the article on the section about evaluation:
> 
> > Just as with instantiation, this is done as a depth first post-order traversal.
> 
> Because for any module before itself is evaluated, any imported modules from top to bottom are evaluated (unless they already were evaluated elsewhere before), the imported module `dotenv` is evaluated first, and then `sub` is evaluated, and oops, `dotenv.config()` didn't run yet. Afterwards, now that all imported modules of `main` were evaluated, `main` itself can be evaluated, and `dotenv.config()` is called, but for the poor `sub` module that's already too late.
> 
> This is solved with `import 'dotenv/config'` because this loads the environment already as part of its own evaluation.

@CherryDT , is this correct though?

I just [tested it](https://github.com/hsuabina/module-import-test) and I get this output:

```
SUB abc
MAIN abc
```

**Comment by crizardevelop** [comment]
Don't try to add a name to the file (like variables.env). Just create a file with the name ".env"

![image](https://github.com/motdotla/dotenv/assets/78008408/3683ee82-d6e8-4ee9-83ad-5c047370639a)

![image](https://github.com/motdotla/dotenv/assets/78008408/f9419aa5-c18e-4715-a56d-6146b3c1feab)

**Comment by pujan789** [comment]
import 'dotenv/config';

**Comment by CherryDT** [comment]
> > @SrBrahma This explains it all very well: https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/
> > Then think about it again - If you have files like this:
> > ```
> > // main.js
> > import dotenv from 'dotenv'
> > dotenv.config()
> > 
> > console.log('MAIN', process.env.STUFF) 
> > 
> > import './sub.js'
> > ```
> > 
> > 
> > ```
> > // sub.js
> > console.log('SUB', process.env.STUFF)
> > ```
> > 
> > 
> > ... and you run `STUFF=abc node main.js`, then you will get this:
> > ```
> > SUB undefined
> > MAIN abc
> > ```
> > 
> > 
> > Remember this sentence from the article on the section about evaluation:
> > > Just as with instantiation, this is done as a depth first post-order traversal.
> > 
> > 
> > Because for any module before itself is evaluated, any imported modules from top to bottom are evaluated (unless they already were evaluated elsewhere before), the imported module `dotenv` is evaluated first, and then `sub` is evaluated, and oops, `dotenv.config()` didn't run yet. Afterwards, now that all imported modules of `main` were evaluated, `main` itself can be evaluated, and `dotenv.config()` is called, but for the poor `sub` module that's already too late.
> > This is solved with `import 'dotenv/config'` because this loads the environment already as part of its own evaluation.
> 
> @CherryDT , is this correct though?
> 
> I just [tested it](https://github.com/hsuabina/module-import-test) and I get this output:
> 
> ```
> SUB abc
> MAIN abc
> ```

You are right, I made a mistake while writing this post.

Of course it works because I wrote to use `STUFF=abc node main.js`, so there isn't even any `.env` file involved!

What I _meant_ to write (and changed now accordingly) was to use `node main.js` and creating a `.env` file with `STUFF=abc` in it.

**Comment by volodalexey** [comment]
Doesn't work for my case. I have `nextjs` and `expressjs` applications alognside and I want to start `expressjs` using `.env.local` from `nextjs`.
content `index.ts`:
```typescript
import dotenv from "dotenv";
dotenv.config({
  path: resolve(process.cwd(), ".env.local"),
});
import { env } from "./env" // <-- throws zod parse error
```
content `env.ts`:
```typescript
import z from "zod";
const configSchema = z.object({
  NEXTAUTH_URL: z.string().url(),
});
export const env = configSchema.parse(process.env);
```
The problem is because `process.env` inside `index.ts` is not the same as `process.env` inside `env.ts`.
Running `index.ts` as `tsx index.ts` ([tsx](https://github.com/esbuild-kit/tsx)).

Any ideas how to solve this?

**Comment by CherryDT** [comment]
That's exactly the scenario I explained why `import dotenv from 'dotenv'; dotenv.config()` doesn't work and `import 'dotenv/config'` has to be used.

In your case you need to pass arguments to `.config` so you can't use that method. Instead, you can move your `dotenv` setup into a separate module and include that in your main module at the very top:

`index.ts`:
```javascript
import './setupDotEnv';
import { env } from "./env";
```

`setupDotEnv.ts`:
```javascript
import dotenv from "dotenv";
dotenv.config({
  path: resolve(process.cwd(), ".env.local"),
});
```

`env.ts`:
```javascript
import z from "zod";
const configSchema = z.object({
  NEXTAUTH_URL: z.string().url(),
});
export const env = configSchema.parse(process.env);
```

**Comment by paraschouhan15** [comment]
import dotenv from 'dotenv'
import express from "express";
import colors from "colors"


dotenv.config();

const app = express();

app.get('/', (req, res) => {
    res.send(
        "<h1> Welcome to my aapp</h1>"
    );
})
//port

const port = process.env.PORT

app.listen(port, () => {
    console.log(`Server running on mode ${process.env.DEV_MODE} port number ${port}`.bgYellow.black)
})
I am getting undefined...in console PORT number AND developmnt mode

**Comment by luiskurihara** [comment]
> does this not work for you?
> 
> ```js
> import 'dotenv/config`
> ```

YOU SAVED ME THANKSSS

**Comment by motdotla** [comment]
for those coming across this we are starting to recommend using `dotenvx`

* runs anywhere (cross-platform)
* multiple enviornments
* encrypted envs

and you are not going to run into these kind of edge case issues somewhat related to how node has developed over time. dotenvx will also be compatible if you are writing deno or bun code.

**Comment by abhinavph21** [comment]
> Doesn't work for my case. I have `nextjs` and `expressjs` applications alognside and I want to start `expressjs` using `.env.local` from `nextjs`. content `index.ts`:
> 
> ```ts
> import dotenv from "dotenv";
> dotenv.config({
>   path: resolve(process.cwd(), ".env.local"),
> });
> import { env } from "./env" // <-- throws zod parse error
> ```
> 
> content `env.ts`:
> 
> ```ts
> import z from "zod";
> const configSchema = z.object({
>   NEXTAUTH_URL: z.string().url(),
> });
> export const env = configSchema.parse(process.env);
> ```
> 
> The problem is because `process.env` inside `index.ts` is not the same as `process.env` inside `env.ts`. Running `index.ts` as `tsx index.ts` ([tsx](https://github.com/esbuild-kit/tsx)).
> 
> Any ideas how to solve this?

read this,
 https://nextjs.org/docs/app/api-reference/next-config-js/env,
 in nextjs, you have to create next.config.js for environment variables

**Comment by jonjieviduya** [comment]
I tried every single code here but it always says "process is not defined"

**Comment by GohilTushar** [comment]
import dotenv from "dotenv";
import path from 'path';
dotenv.config({
  path:path.resolve(<Your path for .env>)
});

**Comment by sakhimpungose** [comment]
This thread pointed me in the right direction, thanks! I tweaked one of the solutions to suit my needs. Here’s what worked for me:

project-root/
project-root/src/index.js
project-root/src/config/setup.js
project-root/src/config/index.js
project-root/.env
project-root/package.json

```javascript
// project-root/src/config/setup.js

import dotenv from 'dotenv';
const { error } = dotenv.config();
if(error){
    throw new Error('Config file not found.');
}
```

```javascript
// project-root/src/config/index.js

const PORT = process.env.PORT || 5000;

// opportunity to do config validations...

const config = {
    port: PORT
};

export default config;
```

```javascript
// project-root/src/index.js

import './config/setup.js';

import config from './config/index.js';

console.log(config.port); // 5000
```

```env
# project-root/.env
PORT=5000
```

```javascript
// project-root/package.json
{
    "type": "module"
    // ...
}
```
